lisa/sut_orchestrator/azure/platform_.py

```diff
@@ -394,19 +400,50 @@ def _prepare_environment(self, environment: Environment, log: Logger) -> bool:
             for req_index, req in enumerate(nodes_requirement):
                 candidate_caps = self._get_allowed_capabilities(req, location, log)
 
+                # filter vm sizes and return two list. 1st is deployable, 2nd is
+                # wait able for released resource.
+                available_capabilities, awaitable_capabilities = self._check_quota(
+                    candidate_caps
                 )
 
-                # TODO: filter out vm sizes, which has no enough quota.
-
                 # sort vm sizes to match
-                candidate_caps = self.get_sorted_vm_sizes(candidate_caps, log)
+                available_capabilities = self.get_sorted_vm_sizes(
+                    available_capabilities, log
+                )
 
                 # match vm sizes by capability or use the predefined vm sizes.
-                candidate_cap = self._get_matched_capability(req, candidate_caps)
+                candidate_cap = self._get_matched_capability(
+                    req, available_capabilities
+                )
                 if candidate_cap:
                     caps[req_index] = candidate_cap
                     cost += candidate_cap.cost
 
+                if not candidate_cap:
+                    # check if there is awaitable VMs
+                    candidate_cap = self._get_matched_capability(
+                        req, awaitable_capabilities
+                    )
+                    if candidate_cap:
+                        # True means awaitable.
+                        caps[req_index] = True
+
+            # Check if sum of the same capabilities over the cap. If so, mark
+            # the overflow cap as True.
+            if all(isinstance(x, schema.NodeSpace) for x in caps):
+                cap_calculator: Dict[str, Tuple[int, int]] = {}
+                for index, cap in enumerate(caps):
+                    assert isinstance(cap, schema.NodeSpace), f"actual: {type(cap)}"
+                    azure_runbook = cap.get_extended_runbook(AzureNodeSchema, AZURE)
+                    vm_size = azure_runbook.vm_size
+                    if vm_size not in cap_calculator:
+                        cap_calculator[vm_size] = self._get_usage(location, vm_size)
+                    remaining, total = cap_calculator[vm_size]
+                    remaining += 1
```

Comment:
Didnt understand this part. Why we are incrementing "remaining and checking if remaining is > total". Can't we just check if "remaining" meets the requirement or not. I think i am missing something here

Reply:
The logic counts remaining by vm_size, and make sure the whole env can be deployed, without quota issue. The "total" is the hard limitation of Azure. I should change the name from "total" to "limits".